Add a new module for akka-http models

himanshu4141 opened the pull request:
This PR adds support for [Akka-Http](https://doc.akka.io/docs/akka-http/current/common/http-model.html#http-model)'s `Uri` .

himanshu4141 committed:
Add a new module for akka-http models

himanshu4141 committed:
Merge branch 'master' into akka-http-module

himanshu4141 committed:
fix formatting

himanshu4141 committed:
Merge branch 'akka-http-module' of github.com:himanshu4141/pureconfig into akka-http-module

himanshu4141 commented:
@ruippeixotog Hello, need a bit of help here. I am not able to figure out why website creation is failing.

jcazevedo commented:
@himanshu4141: The failed check indicates that this PR produces a change to the contents of PureConfig's website. It's expected if you change the documentation.

@ruippeixotog recently changed the CI step so that the build would no longer fail (#792). If you sync your branch with master, the failed check should disappear.

himanshu4141 committed:
Merge branch 'master' into akka-http-module

himanshu4141 commented:
@jcazevedo ahh ok. The website change I guess would be the link I added to include the new module.

Thanks for the quick response. I have updated the branch. Hopefully, this should go away, making this PR ready for review.

ruippeixotog commented on modules/akka-http/README.md:
```diff
@@ -0,0 +1,37 @@
+
+# Akka-Http module for PureConfig
```
Can you please use tut for the documentation? You should move this to `modules/akka-http/src/main/tut/README.md`, use tags as in [this example](https://github.com/pureconfig/pureconfig/blob/master/modules/akka/src/main/tut/README.md) and run `sbt tut`.

himanshu4141 replied:
sure

ruippeixotog commented on modules/akka-http/build.sbt:
```diff
@@ -0,0 +1,16 @@
+crossScalaVersions ~= { _.filterNot(_.startsWith("2.11")) }
+
+libraryDependencies ++= Seq(
+  "com.typesafe.akka" %% "akka-actor" % "2.5.31",
```
Why is akka-actor required here?

himanshu4141 replied:
~It's a required dependency of akka-http.~

Actually latest version relies upon `akka-stream` not `akka-actor` as per developer [docs](https://doc.akka.io/docs/akka-http/current/introduction.html#using-akka-http). I"ll update.

ruippeixotog commented on modules/akka-http/build.sbt:
```diff
@@ -0,0 +1,16 @@
+
+OsgiKeys.exportPackage := Seq("pureconfig.module.akkahttp.*")
+OsgiKeys.privatePackage := Seq()
+OsgiKeys.importPackage := Seq(s"""scala.*;version="[${scalaBinaryVersion.value}.0,${scalaBinaryVersion.value}.50)"""", "*")
```
Can you add a newline at the end?

ruippeixotog commented on modules/akka-http/src/main/scala/pureconfig/module/akkahttp/package.scala:
```diff
@@ -0,0 +1,24 @@
+      Try(Uri(str, ParsingMode.Strict)).fold(
+        {
+          case err: IllegalUriException => Left(CannotConvert(str, "Uri", err.info.summary))
+          case err => Left(CannotConvert(str, "Uri", err.getMessage.replace('^', " ".charAt(0)).trim))
```
Why are we parsing the message here, in the fallback exception case? If we have no information on why an exception was thrown we should use `ExceptionThrown` instead of `CannotConvert`.

himanshu4141 replied:
used `ExceptionThrown` for fallback cases.

ruippeixotog commented on modules/akka-http/src/main/scala/pureconfig/module/akkahttp/package.scala:
```diff
@@ -0,0 +1,24 @@
+
+  implicit val uriReader: ConfigReader[Uri] =
+    ConfigReader.fromString(str =>
+      Try(Uri(str, ParsingMode.Strict)).fold(
```
nit: you can use `Try(...).toEither.left.map(...)` to simplify this expression, by getting rid of the success `case` and the need to wrap errors in `Left`.

ruippeixotog reviewed (changes requested):
Thanks for the contribution, @himanshu4141!

himanshu4141 committed:
review comments -

1. use tut for doc
2. Handle Exception better

ruippeixotog commented on modules/akka-http/build.sbt:
```diff
@@ -0,0 +1,16 @@
+crossScalaVersions ~= { _.filterNot(_.startsWith("2.11")) }
+
+libraryDependencies ++= Seq(
+  "com.typesafe.akka" %% "akka-stream" % "2.5.31",
```
I see. Nonetheless, this being a library project that does not depend explicitly on any code from `akka-stream`, it's still important that we mark `akka-stream` as `provided` in order to make sure we don't bind client code to a specific version (and force them to provide their own akka-stream version as described [here](https://doc.akka.io/docs/akka-http/current/compatibility-guidelines.html#compatibility-with-akka)).
```suggestion
  "com.typesafe.akka" %% "akka-stream" % "2.5.31" % Provided,
```

himanshu4141 replied:
hmm, didn't think of that. Sounds good. Thanks for the suggestion

ruippeixotog replied:
I forgot about tut, which still needs it to run code in the documentation. I took the liberty of adding tut to the dependency scopes myself, hope it's ok :)

ruippeixotog reviewed (changes requested):
LGTM! Please take care of this last comment and we'll merge it.

himanshu4141 committed:
Update modules/akka-http/build.sbt

make akka-stream provided to allow clients specifying their own version.

Co-authored-by: Rui Gonçalves <ruippeixotog@gmail.com>

ruippeixotog committed:
Add tut to akka-stream scopes

ruippeixotog committed:
Merge branch 'master' into akka-http-module

himanshu4141 commented:
@ruippeixotog thanks for the merge and that last change. 👍
